File: 2020/typescript/src/solve.ts
import { readFileSync } from 'fs';

const exec = (fn: Function, input: string) => {
  const start = process.hrtime();
  const result = fn(input);
  const end = process.hrtime(start);

  return { result, duration: (end[0] * 1000000000 + end[1]) / 1000000 };
};

const [, , dayNo, solutionNo = 1] = process.argv;

console.log(`Day ${dayNo}, Solution ${solutionNo}`);

const daySrcDir = `${__dirname}/day${dayNo.toString().padStart(2, '0')}`;
const { part1, part2 } = require(`${daySrcDir}/solution-${solutionNo}`);
const input = readFileSync(`${daySrcDir}/input.txt`, 'utf-8').trim();

const p1 = exec(part1, input);
console.log(`Part 1: ${p1.result} [time taken: ${p1.duration} ms]`);

const p2 = exec(part2, input);
console.log(`Part 2: ${p2.result} [time taken: ${p2.duration} ms]`);

File: 2016/python/src/solve.py
import sys
import importlib.util
import os
import datetime


def solve(part, input_content):
    start_time = datetime.datetime.now()
    answer = part(input_content)
    return answer, (datetime.datetime.now() - start_time).total_seconds() * 1000


def load_solution(file_path):
    spec = importlib.util.spec_from_file_location("solution", file_path)
    solution = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(solution)
    return solution


def read_input(file_path):
    with open(file_path, 'r') as file:
        return file.read()


def main(argv):
    day_no, solution_no = map(int, argv)

    day_dir = "{}/day{:02}".format(os.path.dirname(__file__), day_no)
    solution = load_solution("{}/solution-{}.py".format(day_dir, solution_no))
    input_content = read_input(day_dir + "/input.txt")

    print("Day {}, Solution {}".format(day_no, solution_no))

    if (hasattr(solution, 'part1')):
        answer, duration = solve(solution.part1, input_content)
        print("Part 1: {} [time taken: {:.3f} ms]".format(answer, duration))

    if (hasattr(solution, 'part2')):
        answer, duration = solve(solution.part2, input_content)
        print("Part 2: {} [time taken: {:.3f} ms]".format(answer, duration))


if __name__ == "__main__":
    main(sys.argv[1:])
